src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let copyState = { ...state };
+  let newState = {};
+  const historyState = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        newState = { ...copyState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        newState = { ...copyState };
+        action.keysToRemove.forEach((key) => delete newState[key]);
+        break;
+      case 'clear':
+        newState = {};
+        break;
+      default:
+        newState = copyState;
```

Comment:
The default case should not simply assign `copyState` to `newState`. It is better to throw an error or handle unexpected action types appropriately.